File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/stored/library/items/GivenASetOfStoredItems/AndCollectionIsCancelledAfterStoredItemsAreReturned/WhenCollectingTheAssociatedServiceFiles.java
package com.lasthopesoftware.bluewater.client.stored.library.items.GivenASetOfStoredItems.AndCollectionIsCancelledAfterStoredItemsAreReturned;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import com.annimon.stream.Stream;
import com.lasthopesoftware.bluewater.client.browsing.items.ItemId;
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.access.ProvideLibraryFiles;
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.access.parameters.FileListParameters;
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId;
import com.lasthopesoftware.bluewater.client.stored.library.items.FakeDeferredStoredItemAccess;
import com.lasthopesoftware.bluewater.client.stored.library.items.StoredItem;
import com.lasthopesoftware.bluewater.client.stored.library.items.StoredItemServiceFileCollector;
import com.lasthopesoftware.bluewater.shared.promises.extensions.FuturePromise;
import com.namehillsoftware.handoff.promises.Promise;

import org.junit.BeforeClass;
import org.junit.Test;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Random;
import java.util.concurrent.CancellationException;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class WhenCollectingTheAssociatedServiceFiles {

	private static final List<ServiceFile> firstItemExpectedFiles = givenARandomCollectionOfFiles();
	private static final List<ServiceFile> secondItemExpectedFiles = givenARandomCollectionOfFiles();
	private static final List<ServiceFile> thirdItemExpectedFiles = givenARandomCollectionOfFiles();
	private static Throwable exception;

	@BeforeClass
	public static void before() throws InterruptedException, TimeoutException {

		final FakeDeferredStoredItemAccess storedItemAccess = new FakeDeferredStoredItemAccess() {
			@Override
			protected Collection<StoredItem> getStoredItems() {
				return Arrays.asList(
					new StoredItem(1, 1, StoredItem.ItemType.ITEM),
					new StoredItem(1, 2, StoredItem.ItemType.ITEM),
					new StoredItem(1, 3, StoredItem.ItemType.ITEM));
			}
		};

		final FileListParameters fileListParameters = FileListParameters.INSTANCE;

		final ProvideLibraryFiles fileProvider = mock(ProvideLibraryFiles.class);
		when(fileProvider.promiseFiles(new LibraryId(2), FileListParameters.Options.None, fileListParameters.getFileListParameters(new ItemId(1))))
			.thenReturn(new Promise<>(firstItemExpectedFiles));
		when(fileProvider.promiseFiles(new LibraryId(2), FileListParameters.Options.None, fileListParameters.getFileListParameters(new ItemId(2))))
			.thenReturn(new Promise<>(secondItemExpectedFiles));
		when(fileProvider.promiseFiles(new LibraryId(2), FileListParameters.Options.None, fileListParameters.getFileListParameters(new ItemId(3))))
			.thenReturn(new Promise<>(thirdItemExpectedFiles));

		final StoredItemServiceFileCollector serviceFileCollector = new StoredItemServiceFileCollector(
			storedItemAccess,
			fileProvider,
			fileListParameters);

		final Promise<Collection<ServiceFile>> serviceFilesPromise = serviceFileCollector.promiseServiceFilesToSync(new LibraryId(2));

		serviceFilesPromise.cancel();

		storedItemAccess.resolveStoredItems();

		try {
			new FuturePromise<>(serviceFilesPromise).get(1, TimeUnit.SECONDS);
		} catch (ExecutionException e) {
			exception = e.getCause();
		}
	}

	@Test
	public void thenACancellationExceptionIsThrown() {
		assertThat(exception).isInstanceOf(CancellationException.class);
	}

	private static List<ServiceFile> givenARandomCollectionOfFiles() {
		final Random random = new Random();
		final int floor = random.nextInt(10000);
		final int ceiling = random.nextInt(10000 - floor) + floor;
		return Stream.range(floor, ceiling).map(ServiceFile::new).toList();
	}

}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/stored/library/items/GivenASetOfStoredItems/AndCollectionIsCancelledAfterStoredItemsAreReturned/WhenCollectingTheAssociatedServiceFiles.kt
package com.lasthopesoftware.bluewater.client.stored.library.items.GivenASetOfStoredItems.AndCollectionIsCancelledAfterStoredItemsAreReturned

import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.files.access.ProvideLibraryFiles
import com.lasthopesoftware.bluewater.client.browsing.items.ItemId
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.stored.library.items.FakeDeferredStoredItemAccess
import com.lasthopesoftware.bluewater.client.stored.library.items.StoredItem
import com.lasthopesoftware.bluewater.client.stored.library.items.StoredItemServiceFileCollector
import com.lasthopesoftware.bluewater.shared.promises.extensions.ExpiringFuturePromise
import com.lasthopesoftware.promises.extensions.toPromise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import java.util.Random
import java.util.concurrent.CancellationException
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeUnit

class WhenCollectingTheAssociatedServiceFiles {
	private val firstItemExpectedFiles = givenARandomCollectionOfFiles()
	private val secondItemExpectedFiles = givenARandomCollectionOfFiles()
	private val thirdItemExpectedFiles = givenARandomCollectionOfFiles()
	private var exception: Throwable? = null

	@BeforeAll
	fun before() {
		val storedItemAccess: FakeDeferredStoredItemAccess =
			object : FakeDeferredStoredItemAccess() {
				override val storedItems: Collection<StoredItem>
					get() = listOf(
						StoredItem(1, "1", StoredItem.ItemType.ITEM),
						StoredItem(1, "2", StoredItem.ItemType.ITEM),
						StoredItem(1, "3", StoredItem.ItemType.ITEM)
					)
			}

		val fileProvider = mockk<ProvideLibraryFiles> {
			every {
				promiseFiles(LibraryId(2), ItemId("1"))
			} returns firstItemExpectedFiles.toPromise()

			every {
				promiseFiles(LibraryId(2), ItemId("2"))
			} returns secondItemExpectedFiles.toPromise()

			every {
				promiseFiles(LibraryId(2), ItemId("3"))
			} returns thirdItemExpectedFiles.toPromise()
		}

		val serviceFileCollector = StoredItemServiceFileCollector(
            storedItemAccess,
            fileProvider
        )

		val serviceFilesPromise = serviceFileCollector.promiseServiceFilesToSync(LibraryId(2))
		serviceFilesPromise.cancel()
		storedItemAccess.resolveStoredItems()
		try {
			ExpiringFuturePromise(serviceFilesPromise)[1, TimeUnit.SECONDS]
		} catch (e: ExecutionException) {
			exception = e.cause
		}
	}

	@Test
	fun `then a cancellation exception is thrown`() {
		assertThat(exception).isInstanceOf(CancellationException::class.java)
	}

	companion object {
		private fun givenARandomCollectionOfFiles(): List<ServiceFile> {
			val random = Random()
			val floor = random.nextInt(10000)
			val ceiling = random.nextInt(10000 - floor) + floor
			return (floor..ceiling).map { ServiceFile(it.toString()) }
		}
	}
}
